pkg/login/social/okta_oauth.go

```diff
@@ -0,0 +1,151 @@
+	Name        string              `json:"name"`
+	DisplayName string              `json:"display_name"`
+	Login       string              `json:"login"`
+	Username    string              `json:"username"`
+	Email       string              `json:"email"`
+	Upn         string              `json:"upn"`
+	Attributes  map[string][]string `json:"attributes"`
+	Groups      []string            `json:"groups"`
+	rawJSON     []byte
+}
+
+type OktaClaims struct {
+	ID                string `json:"sub"`
+	Email             string `json:"email"`
+	PreferredUsername string `json:"preferred_username"`
+	Name              string `json:"name"`
+}
+
+func (claims *OktaClaims) extractEmail() string {
+	if claims.Email == "" && claims.PreferredUsername != "" {
+		return claims.PreferredUsername
+	}
+
+	return claims.Email
+}
+
+func (s *SocialOkta) Type() int {
+	return int(models.OKTA)
+}
+
+func (s *SocialOkta) UserInfo(client *http.Client, token *oauth2.Token) (*BasicUserInfo, error) {
+	idToken := token.Extra("id_token")
+	if idToken == nil {
+		return nil, fmt.Errorf("No id_token found")
+	}
+
+	parsedToken, err := jwt.ParseSigned(idToken.(string))
+	if err != nil {
+		return nil, errutil.Wrapf(err, "error parsing id token")
+	}
+
+	var claims OktaClaims
+	if err := parsedToken.UnsafeClaimsWithoutVerification(&claims); err != nil {
+		return nil, errutil.Wrapf(err, "error getting claims from id token")
+	}
+
+	email := claims.extractEmail()
+	if email == "" {
+		return nil, errors.New("error getting user info: no email found in access token")
+	}
+
+	var data OktaUserInfoJson
+	err = s.extractAPI(&data, client)
+	if err != nil {
+		s.log.Error("Error extracting user info", "error", err)
+	}
+
+	role := s.extractRole(&data)
+
+	groups := s.GetGroups(&data)
+	if !s.IsGroupMember(groups) {
+		return nil, ErrMissingGroupMembership
+	}
+
+	return &BasicUserInfo{
+		Id:     claims.ID,
+		Name:   claims.Name,
+		Email:  email,
+		Login:  email,
+		Role:   role,
+		Groups: groups,
+	}, nil
+}
+
+func (s *SocialOkta) extractAPI(data *OktaUserInfoJson, client *http.Client) error {
+	rawUserInfoResponse, err := HttpGet(client, s.apiUrl)
+	if err != nil {
+		s.log.Debug("Error getting user info response", "url", s.apiUrl, "error", err)
+		return errutil.Wrapf(err, "error getting user info response")
+	}
+	data.rawJSON = rawUserInfoResponse.Body
+
+	err = json.Unmarshal(data.rawJSON, data)
+	if err != nil {
+		s.log.Debug("Error decoding user info response", "raw_json", data.rawJSON, "error", err)
+		data.rawJSON = []byte{}
+		return errutil.Wrapf(err, "error decoding user info response")
+	}
+
+	s.log.Debug("Received user info response", "raw_json", string(data.rawJSON), "data", data)
+	return nil
+}
+
+func (s *SocialOkta) extractRole(data *OktaUserInfoJson) string {
+	if s.roleAttributePath == "" {
+		return ""
+	}
+
+	role, err := s.searchJSONForAttr(s.roleAttributePath, data.rawJSON)
+	if err != nil {
+		s.log.Error("Failed to extract role", "error", err)
```

Comment:
I would modify `extractRole` into returning an error, and propagate the error here instead. Is there a reason to swallow the error here?